compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -695,14 +694,19 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     ) -> bool {
         // Reject any attempt to unify two unevaluated constants that contain inference
         // variables, since inference variables in queries lead to ICEs.
-        if a.substs.has_infer_types_or_consts() || b.substs.has_infer_types_or_consts() {
-            debug!("a or b contain infer vars in its substs -> cannot unify");
+        if a.substs.has_infer_types_or_consts()
+            || b.substs.has_infer_types_or_consts()
+            || param_env.has_infer_types_or_consts()
+        {
+            debug!("a or b or param_env contain infer vars in its substs -> cannot unify");
             return false;
         }
 
-        let canonical = self.canonicalize_query((a, b), &mut OriginalQueryValues::default());
+        let erased_args = self.tcx.erase_regions((a, b));
+        let erased_param_env = self.tcx.erase_regions(param_env);
```

Comment:
first use `param_env.and`, then erase regions :laughing: doesn't matter too much but is slightly clearer imo